test/hotspot/gtest/opto/test_constant_division.cpp

```diff
@@ -116,16 +93,56 @@ static void test_division(T d, T N_neg, T N_pos, juint min_s) {
   }
 }
 
+template <class UT, class U>
+static void test_division(UT d, UT N_neg, UT N_pos, juint min_s) {
+  constexpr juint W = sizeof(UT) * 8;
+
+  if ((N_neg < d && N_pos < d) || (d & (d - 1)) == 0) {
+    return;
+  }
+
+  UT c;
+  bool c_ovf;
+  juint s;
+  magic_divide_constants(d, N_neg, N_pos, min_s, c, c_ovf, s);
+
+  auto op = [&](U l) -> U {
+    if (!c_ovf) {
+      return ((l * U(c)) >> s) + (l < 0 ? U(1) : U(0));
+    } else {
+      if (sizeof(U) > sizeof(UT) * 2) {
+        constexpr U wrap_amount = U(UT(-1)) + 1;
+        return (l * (U(c) + wrap_amount)) >> s;
+      } else {
+        U mul_hi = (l * U(c)) >> W;
+        return (((l - mul_hi) >> 1) + mul_hi) >> (s - 1 - W);
+      }
+    }
+  };
+
+  test_op<UT, U>(d, N_neg, N_pos, op);
+}
+
 template <class T, class U>
 static void test_division_random() {
-  constexpr int ITER = 10000;
+  constexpr int iter_num = 10000;
   using UT = std::conditional_t<std::is_same<T, jlong>::value, julong, std::make_unsigned_t<T>>;
-  for (int i = 0; i < ITER; i++) {
+  for (int i = 0; i < iter_num;) {
     UT d = random<T, UT>();
+    if ((d & (d - 1)) == 0) {
+      continue;
+    }
+
     UT N_neg = std::is_signed<T>::value ? random<T, UT>() + 1 : 0;
     UT N_pos = random<T, UT>();
+    if (N_neg < d && N_pos < d) {
+      continue;
```

Comment:
With sufficiently bad luck, we could spin here for a long time.  (Similarly, though much less likely above with
the power-of-2 case.) That doesn't seem great.  Of course, if one does count these skipped cases against
the iteration limit then with sufficiently bad luck one might not test anything.  Rather than skipping the test
here, could you instead modify one of the values and proceed with the test?

Reply:
Yes I have done that, thanks a lot for your suggestions.